File: core/slot_calendar.zig
/// slot_calendar.zig — Pre-computed Slot Calendar (Solana PoH-style)
///
/// Computes and caches the next 60 future slots so that:
///   1. Frontend can display "next leader: ob1qzhrauq0x in 2.3s"
///   2. Future-block pool can route TXs with target_slot=N
///   3. Anti-fork: validators know deterministically who delivers each slot
///
/// Design constraints:
///   - No malloc in SlotCalendar itself — all fixed-size arrays
///   - toJson() takes an allocator (for RPC output only)
///   - No coupling to validator_registry.zig — caller passes validator_set
///     as []const [20]u8 (address bytes).  Hex encoding for RPC is done in
///     toJson().
///
/// Leader election formula (per spec):
///   sha256( slot_id_le8 ++ tip_hash_32 ) → first 8 bytes as u64 → mod N
///
/// NOTE: The existing orchestrator.zig SlotCalendar uses string addresses and
/// a generics-based leaderFn. This module is the self-contained calendar that
/// the spec requested, with [20]u8 address bytes, finalizeSlot, pruneMissed,
/// and a toJson serializer. Both coexist; they serve different purposes.

const std = @import("std");

// ─── Types ──────────────────────────────────────────────────────────────────

pub const SlotState = enum(u8) {
    future = 0,
    in_flight = 1,
    finalized = 2,
    missed = 3,
};

pub const SlotEntry = struct {
    slot_id: u64,
    /// ob1q address bytes — all zeros means no leader (empty validator set).
    leader_address: [20]u8,
    /// Absolute timestamp (ms) when the block is expected to arrive.
    expected_arrival_ms: i64,
    /// SHA-256 of the block that landed — zero-filled until finalized.
    placeholder_hash: [32]u8,
    state: SlotState,
};

pub const MAX_CALENDAR_SLOTS: usize = 60;

// ─── SlotCalendar ────────────────────────────────────────────────────────────

pub const SlotCalendar = struct {
    entries: [MAX_CALENDAR_SLOTS]SlotEntry,
    /// Ring buffer head index — index of the oldest / lowest slot_id entry.
    head: usize,
    /// Chain height at last recompute.
    tip_height: u64,
    /// Chain tip hash (32 bytes) at last recompute.
    tip_hash: [32]u8,
    /// Wall-clock ms when the calendar was last computed.
    computed_at_ms: i64,

    // ─── init ──────────────────────────────────────────────────────────────

    pub fn init() SlotCalendar {
        var cal = SlotCalendar{
            .entries = undefined,
            .head = 0,
            .tip_height = 0,
            .tip_hash = std.mem.zeroes([32]u8),
            .computed_at_ms = 0,
        };
        // Zero all entries so any uncomputed slot is well-defined.
        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            cal.entries[i] = SlotEntry{
                .slot_id = 0,
                .leader_address = std.mem.zeroes([20]u8),
                .expected_arrival_ms = 0,
                .placeholder_hash = std.mem.zeroes([32]u8),
                .state = .future,
            };
        }
        return cal;
    }

    // ─── recompute ─────────────────────────────────────────────────────────

    /// Recompute all 60 future slots from the current chain tip.
    /// Called: at startup, after each finalized block, after governance TX.
    ///
    /// validator_set: slice of 20-byte address arrays. May be empty —
    /// in that case leader_address is all zeros.
    pub fn recompute(
        self: *SlotCalendar,
        tip_height: u64,
        tip_hash: [32]u8,
        now_ms: i64,
        validator_set: []const [20]u8,
    ) void {
        self.tip_height = tip_height;
        self.tip_hash = tip_hash;
        self.computed_at_ms = now_ms;
        self.head = 0; // ring buffer always starts at index 0 after full recompute

        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            const slot_id: u64 = tip_height + 1 + i;
            const arrival_ms: i64 = now_ms + @as(i64, @intCast(i + 1)) * 1000;
            const leader = leaderForSlot(slot_id, tip_hash, validator_set);
            self.entries[i] = SlotEntry{
                .slot_id = slot_id,
                .leader_address = leader,
                .expected_arrival_ms = arrival_ms,
                .placeholder_hash = std.mem.zeroes([32]u8),
                .state = .future,
            };
        }
    }

    // ─── getSlot ───────────────────────────────────────────────────────────

    /// Get the entry for a given slot_id. Returns null if not in window.
    pub fn getSlot(self: *const SlotCalendar, slot_id: u64) ?*const SlotEntry {
        // Linear scan — only 60 entries, negligible cost.
        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            if (self.entries[i].slot_id == slot_id) return &self.entries[i];
        }
        return null;
    }

    // ─── finalizeSlot ──────────────────────────────────────────────────────

    /// Mark a slot as finalized when its block arrives.
    /// Advances the ring buffer head if the slot was at head.
    pub fn finalizeSlot(self: *SlotCalendar, slot_id: u64, block_hash: [32]u8) void {
        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            if (self.entries[i].slot_id == slot_id) {
                self.entries[i].placeholder_hash = block_hash;
                self.entries[i].state = .finalized;
                // Advance head past consecutive finalized entries.
                while (self.head < MAX_CALENDAR_SLOTS and
                    self.entries[self.head].state == .finalized)
                {
                    self.head += 1;
                    if (self.head >= MAX_CALENDAR_SLOTS) self.head = 0;
                }
                return;
            }
        }
    }

    // ─── pruneMissed ───────────────────────────────────────────────────────

    /// Mark expired slots as missed (called every tick / per block).
    /// A slot is missed when now_ms > expected_arrival_ms + 2× slot interval.
    pub fn pruneMissed(self: *SlotCalendar, now_ms: i64) void {
        const SLOT_INTERVAL_MS: i64 = 1000;
        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            const e = &self.entries[i];
            if (e.state == .future or e.state == .in_flight) {
                const overdue = now_ms - e.expected_arrival_ms;
                if (overdue >= SLOT_INTERVAL_MS * 2) {
                    e.state = .missed;
                } else if (overdue >= 0) {
                    e.state = .in_flight;
                }
            }
        }
    }

    // ─── toJson ────────────────────────────────────────────────────────────

    /// Serialize the full calendar to JSON for RPC output.
    /// Returns a heap-allocated slice owned by the caller.
    pub fn toJson(self: *const SlotCalendar, alloc: std.mem.Allocator) ![]u8 {
        var buf = std.array_list.Managed(u8).init(alloc);
        errdefer buf.deinit();
        const w = buf.writer();

        try w.writeAll("{\"tip_height\":");
        try w.print("{d}", .{self.tip_height});
        try w.writeAll(",\"computed_at_ms\":");
        try w.print("{d}", .{self.computed_at_ms});
        try w.writeAll(",\"entries\":[");

        var i: usize = 0;
        while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
            if (i > 0) try w.writeAll(",");
            const e = &self.entries[i];
            const state_str: []const u8 = switch (e.state) {
                .future => "future",
                .in_flight => "in_flight",
                .finalized => "finalized",
                .missed => "missed",
            };
            // Encode leader_address as lowercase hex.
            var leader_hex: [40]u8 = undefined;
            _ = std.fmt.bufPrint(&leader_hex, "{s}", .{
                std.fmt.bytesToHex(e.leader_address, .lower),
            }) catch unreachable;
            // Encode placeholder_hash as lowercase hex.
            var hash_hex: [64]u8 = undefined;
            _ = std.fmt.bufPrint(&hash_hex, "{s}", .{
                std.fmt.bytesToHex(e.placeholder_hash, .lower),
            }) catch unreachable;

            try w.print(
                "{{\"slot_id\":{d},\"leader\":\"{s}\",\"expected_arrival_ms\":{d},\"placeholder_hash\":\"{s}\",\"state\":\"{s}\"}}",
                .{ e.slot_id, leader_hex, e.expected_arrival_ms, hash_hex, state_str },
            );
        }

        try w.writeAll("]}");
        return buf.toOwnedSlice();
    }
};

// ─── leaderForSlot ──────────────────────────────────────────────────────────

/// Deterministic leader election (pure function, no state).
///
/// Algorithm (per spec):
///   sha256( slot_id_le8 ++ tip_hash_32 )[0..8] as u64 → mod validator_set.len
///
/// Returns all-zeros if validator_set is empty.
pub fn leaderForSlot(
    slot_id: u64,
    tip_hash: [32]u8,
    validator_set: []const [20]u8,
) [20]u8 {
    if (validator_set.len == 0) return std.mem.zeroes([20]u8);

    var hasher = std.crypto.hash.sha2.Sha256.init(.{});
    var slot_bytes: [8]u8 = undefined;
    std.mem.writeInt(u64, &slot_bytes, slot_id, .little);
    hasher.update(&slot_bytes);
    hasher.update(&tip_hash);
    var digest: [32]u8 = undefined;
    hasher.final(&digest);

    const idx_u64 = std.mem.readInt(u64, digest[0..8], .little);
    const idx: usize = @intCast(idx_u64 % @as(u64, @intCast(validator_set.len)));
    return validator_set[idx];
}

// ─── Tests ───────────────────────────────────────────────────────────────────

const testing = std.testing;

// ── helper: build a small validator set ──────────────────────────────────────
fn makeValidatorSet(n: usize, out: *[8][20]u8) []const [20]u8 {
    var i: usize = 0;
    while (i < n and i < 8) : (i += 1) {
        var addr = std.mem.zeroes([20]u8);
        addr[0] = @intCast(i + 1); // distinct addresses
        out[i] = addr;
    }
    return out[0..n];
}

test "1. init() produces 60 entries all with state=future" {
    const cal = SlotCalendar.init();
    try testing.expectEqual(@as(usize, 0), cal.head);
    var i: usize = 0;
    while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
        try testing.expectEqual(SlotState.future, cal.entries[i].state);
    }
}

test "2. recompute() sets expected_arrival_ms correctly" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(2, &vs_buf);
    const tip_hash = std.mem.zeroes([32]u8);
    const now_ms: i64 = 10_000_000;

    cal.recompute(100, tip_hash, now_ms, vs);

    try testing.expectEqual(@as(usize, 60), MAX_CALENDAR_SLOTS);
    // slot N+1 = now + 1000
    try testing.expectEqual(@as(i64, now_ms + 1000), cal.entries[0].expected_arrival_ms);
    // slot N+2 = now + 2000
    try testing.expectEqual(@as(i64, now_ms + 2000), cal.entries[1].expected_arrival_ms);
    // slot N+60 = now + 60000
    try testing.expectEqual(@as(i64, now_ms + 60_000), cal.entries[59].expected_arrival_ms);
    // slot ids
    try testing.expectEqual(@as(u64, 101), cal.entries[0].slot_id);
    try testing.expectEqual(@as(u64, 160), cal.entries[59].slot_id);
}

test "3. leaderForSlot is deterministic" {
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(3, &vs_buf);
    const tip_hash = [32]u8{ 0xde, 0xad, 0xbe, 0xef, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };

    const r1 = leaderForSlot(42, tip_hash, vs);
    const r2 = leaderForSlot(42, tip_hash, vs);
    try testing.expectEqualSlices(u8, &r1, &r2);
}

test "4. leaderForSlot distributes across validator set" {
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(4, &vs_buf);
    const tip_hash = std.mem.zeroes([32]u8);

    // Count how many of the 60 slots go to each validator.
    var counts = [4]u32{ 0, 0, 0, 0 };
    var slot: u64 = 1;
    while (slot <= 60) : (slot += 1) {
        const leader = leaderForSlot(slot, tip_hash, vs);
        var vi: usize = 0;
        while (vi < 4) : (vi += 1) {
            if (std.mem.eql(u8, &leader, &vs[vi])) {
                counts[vi] += 1;
                break;
            }
        }
    }
    // With SHA256 distribution each validator should get roughly 15 slots.
    // Allow generous margin (5–35) to avoid flakiness without being useless.
    for (counts) |c| {
        try testing.expect(c >= 5);
        try testing.expect(c <= 35);
    }
}

test "5. finalizeSlot changes state to finalized and stores hash" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(1, &vs_buf);
    cal.recompute(10, std.mem.zeroes([32]u8), 0, vs);

    var block_hash: [32]u8 = undefined;
    @memset(&block_hash, 0xAB);
    cal.finalizeSlot(11, block_hash); // slot_id = tip+1

    try testing.expectEqual(SlotState.finalized, cal.entries[0].state);
    try testing.expectEqualSlices(u8, &block_hash, &cal.entries[0].placeholder_hash);
}

test "6. pruneMissed marks expired slots as missed" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(1, &vs_buf);
    const now_ms: i64 = 1_000_000;
    cal.recompute(0, std.mem.zeroes([32]u8), now_ms, vs);

    // entry[0] expected at now+1000=1_001_000
    // entry[1] expected at now+2000=1_002_000
    // Jump to now+5000 → entry[0] is overdue by 4s (4× interval) → missed
    //                     entry[1] is overdue by 3s (3× interval) → missed
    //                     entry[2] expected at 1_003_000, overdue 2s (2× interval) → missed
    //                     entry[3] expected at 1_004_000, overdue 1s (< 2×) → in_flight
    cal.pruneMissed(now_ms + 5_000);

    try testing.expectEqual(SlotState.missed, cal.entries[0].state);
    try testing.expectEqual(SlotState.missed, cal.entries[1].state);
    try testing.expectEqual(SlotState.missed, cal.entries[2].state);
    try testing.expectEqual(SlotState.in_flight, cal.entries[3].state);
    // entry[4] expected at 1_005_000 — not yet overdue (now == expected) → in_flight
    try testing.expectEqual(SlotState.in_flight, cal.entries[4].state);
    // entry[5] expected at 1_006_000 — still in the future
    try testing.expectEqual(SlotState.future, cal.entries[5].state);
}

test "7. getSlot returns null for slot outside window" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(1, &vs_buf);
    cal.recompute(100, std.mem.zeroes([32]u8), 0, vs);

    // Slots 101–160 are in window.
    try testing.expect(cal.getSlot(101) != null);
    try testing.expect(cal.getSlot(160) != null);
    // Slot 100 (tip) and 161 (outside) should not be found.
    try testing.expect(cal.getSlot(100) == null);
    try testing.expect(cal.getSlot(161) == null);
    // Slot 0 is definitely not in window.
    try testing.expect(cal.getSlot(0) == null);
}

test "8. recompute with empty validator_set gives zero leader_address" {
    var cal = SlotCalendar.init();
    const vs: []const [20]u8 = &[_][20]u8{};
    cal.recompute(0, std.mem.zeroes([32]u8), 0, vs);

    const zero_addr = std.mem.zeroes([20]u8);
    var i: usize = 0;
    while (i < MAX_CALENDAR_SLOTS) : (i += 1) {
        try testing.expectEqualSlices(u8, &zero_addr, &cal.entries[i].leader_address);
    }
}

test "9. ring buffer head advances past consecutive finalized entries" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(1, &vs_buf);
    cal.recompute(0, std.mem.zeroes([32]u8), 0, vs);

    try testing.expectEqual(@as(usize, 0), cal.head);

    // Finalize slots 1, 2, 3 in order.
    var hash: [32]u8 = undefined;
    @memset(&hash, 0x01);
    cal.finalizeSlot(1, hash);
    // head should advance past slot_id=1 (entries[0])
    try testing.expectEqual(@as(usize, 1), cal.head);

    @memset(&hash, 0x02);
    cal.finalizeSlot(2, hash);
    try testing.expectEqual(@as(usize, 2), cal.head);

    // Finalizing slot 4 (not consecutive with head=2 → entry[2] is slot 3)
    // shouldn't advance head because entries[2] (slot 3) is still future.
    @memset(&hash, 0x04);
    cal.finalizeSlot(4, hash);
    // head remains at 2 because entries[2] (slot 3) is not yet finalized.
    try testing.expectEqual(@as(usize, 2), cal.head);
}

test "10. toJson produces valid parsable JSON" {
    var cal = SlotCalendar.init();
    var vs_buf: [8][20]u8 = undefined;
    const vs = makeValidatorSet(2, &vs_buf);
    cal.recompute(5, std.mem.zeroes([32]u8), 12345, vs);

    const json = try cal.toJson(testing.allocator);
    defer testing.allocator.free(json);

    // Must contain key fields.
    try testing.expect(std.mem.indexOf(u8, json, "\"entries\"") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"slot_id\"") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"state\"") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"future\"") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"tip_height\":5") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"computed_at_ms\":12345") != null);
    // Must be parseable by std.json.
    const parsed = try std.json.parseFromSlice(std.json.Value, testing.allocator, json, .{});
    defer parsed.deinit();
    const obj = parsed.value.object;
    try testing.expect(obj.contains("entries"));
}

File: core-rust/src/consensus/slot_calendar.rs
//! Pre-computed Slot Calendar (Solana PoH-style) — port of `core/slot_calendar.zig`.
//!
//! Computes and caches the next 60 future slots so that:
//!   1. Frontend can display "next leader: ob1qzhrauq0x in 2.3s"
//!   2. Future-block pool can route TXs with target_slot=N
//!   3. Anti-fork: validators know deterministically who delivers each slot
//!
//! Leader election: sha256(slot_id_le8 || tip_hash_32)[0..8] as u64 → mod N.

use sha2::{Digest, Sha256};

pub const MAX_CALENDAR_SLOTS: usize = 60;
pub const SLOT_INTERVAL_MS: i64 = 1000;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum SlotState {
    Future = 0,
    InFlight = 1,
    Finalized = 2,
    Missed = 3,
}

impl SlotState {
    pub fn as_str(self) -> &'static str {
        match self {
            SlotState::Future => "future",
            SlotState::InFlight => "in_flight",
            SlotState::Finalized => "finalized",
            SlotState::Missed => "missed",
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct SlotEntry {
    pub slot_id: u64,
    /// ob1q address bytes — all zeros means no leader (empty validator set).
    pub leader_address: [u8; 20],
    /// Absolute timestamp (ms) when the block is expected to arrive.
    pub expected_arrival_ms: i64,
    /// SHA-256 of the block that landed — zero-filled until finalized.
    pub placeholder_hash: [u8; 32],
    pub state: SlotState,
}

impl Default for SlotEntry {
    fn default() -> Self {
        Self {
            slot_id: 0,
            leader_address: [0u8; 20],
            expected_arrival_ms: 0,
            placeholder_hash: [0u8; 32],
            state: SlotState::Future,
        }
    }
}

pub struct SlotCalendar {
    pub entries: [SlotEntry; MAX_CALENDAR_SLOTS],
    /// Ring buffer head — index of the oldest non-finalized slot.
    pub head: usize,
    pub tip_height: u64,
    pub tip_hash: [u8; 32],
    pub computed_at_ms: i64,
}

impl SlotCalendar {
    pub fn new() -> Self {
        Self {
            entries: [SlotEntry::default(); MAX_CALENDAR_SLOTS],
            head: 0,
            tip_height: 0,
            tip_hash: [0u8; 32],
            computed_at_ms: 0,
        }
    }

    /// Recompute all 60 future slots from the current chain tip.
    pub fn recompute(
        &mut self,
        tip_height: u64,
        tip_hash: [u8; 32],
        now_ms: i64,
        validator_set: &[[u8; 20]],
    ) {
        self.tip_height = tip_height;
        self.tip_hash = tip_hash;
        self.computed_at_ms = now_ms;
        self.head = 0;

        for i in 0..MAX_CALENDAR_SLOTS {
            let slot_id = tip_height + 1 + i as u64;
            let arrival_ms = now_ms + (i as i64 + 1) * SLOT_INTERVAL_MS;
            self.entries[i] = SlotEntry {
                slot_id,
                leader_address: leader_for_slot(slot_id, tip_hash, validator_set),
                expected_arrival_ms: arrival_ms,
                placeholder_hash: [0u8; 32],
                state: SlotState::Future,
            };
        }
    }

    /// Get the entry for a given slot_id (linear scan over 60 entries).
    pub fn get_slot(&self, slot_id: u64) -> Option<&SlotEntry> {
        self.entries.iter().find(|e| e.slot_id == slot_id)
    }

    /// Mark a slot as finalized when its block arrives. Advances head past
    /// consecutive finalized entries.
    pub fn finalize_slot(&mut self, slot_id: u64, block_hash: [u8; 32]) {
        for i in 0..MAX_CALENDAR_SLOTS {
            if self.entries[i].slot_id == slot_id {
                self.entries[i].placeholder_hash = block_hash;
                self.entries[i].state = SlotState::Finalized;
                while self.head < MAX_CALENDAR_SLOTS
                    && self.entries[self.head].state == SlotState::Finalized
                {
                    self.head += 1;
                    if self.head >= MAX_CALENDAR_SLOTS {
                        self.head = 0;
                        break;
                    }
                }
                return;
            }
        }
    }

    /// Mark expired slots as missed; in-window-but-past as in_flight.
    pub fn prune_missed(&mut self, now_ms: i64) {
        for e in self.entries.iter_mut() {
            if e.state == SlotState::Future || e.state == SlotState::InFlight {
                let overdue = now_ms - e.expected_arrival_ms;
                if overdue >= SLOT_INTERVAL_MS * 2 {
                    e.state = SlotState::Missed;
                } else if overdue >= 0 {
                    e.state = SlotState::InFlight;
                }
            }
        }
    }

    /// Serialize the full calendar to JSON for RPC output.
    pub fn to_json(&self) -> String {
        let mut out = String::with_capacity(8192);
        out.push_str("{\"tip_height\":");
        out.push_str(&self.tip_height.to_string());
        out.push_str(",\"computed_at_ms\":");
        out.push_str(&self.computed_at_ms.to_string());
        out.push_str(",\"entries\":[");
        for (i, e) in self.entries.iter().enumerate() {
            if i > 0 {
                out.push(',');
            }
            out.push_str("{\"slot_id\":");
            out.push_str(&e.slot_id.to_string());
            out.push_str(",\"leader\":\"");
            out.push_str(&hex::encode(e.leader_address));
            out.push_str("\",\"expected_arrival_ms\":");
            out.push_str(&e.expected_arrival_ms.to_string());
            out.push_str(",\"placeholder_hash\":\"");
            out.push_str(&hex::encode(e.placeholder_hash));
            out.push_str("\",\"state\":\"");
            out.push_str(e.state.as_str());
            out.push_str("\"}");
        }
        out.push_str("]}");
        out
    }
}

impl Default for SlotCalendar {
    fn default() -> Self {
        Self::new()
    }
}

/// Deterministic leader election (pure function, no state).
///
/// Algorithm: sha256(slot_id_le8 || tip_hash_32)[0..8] as u64 → mod N.
/// Returns all-zeros if validator_set is empty.
pub fn leader_for_slot(
    slot_id: u64,
    tip_hash: [u8; 32],
    validator_set: &[[u8; 20]],
) -> [u8; 20] {
    if validator_set.is_empty() {
        return [0u8; 20];
    }
    let mut hasher = Sha256::new();
    hasher.update(slot_id.to_le_bytes());
    hasher.update(tip_hash);
    let digest = hasher.finalize();
    let idx_u64 = u64::from_le_bytes(digest[0..8].try_into().unwrap());
    let idx = (idx_u64 % validator_set.len() as u64) as usize;
    validator_set[idx]
}

// ─── Tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn make_validator_set(n: usize) -> Vec<[u8; 20]> {
        (0..n)
            .map(|i| {
                let mut a = [0u8; 20];
                a[0] = (i as u8) + 1;
                a
            })
            .collect()
    }

    #[test]
    fn init_produces_60_future_entries() {
        let cal = SlotCalendar::new();
        assert_eq!(cal.head, 0);
        for e in cal.entries.iter() {
            assert_eq!(e.state, SlotState::Future);
        }
    }

    #[test]
    fn recompute_sets_arrival_ms_and_slot_ids() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(2);
        cal.recompute(100, [0u8; 32], 10_000_000, &vs);
        assert_eq!(cal.entries[0].expected_arrival_ms, 10_001_000);
        assert_eq!(cal.entries[1].expected_arrival_ms, 10_002_000);
        assert_eq!(cal.entries[59].expected_arrival_ms, 10_060_000);
        assert_eq!(cal.entries[0].slot_id, 101);
        assert_eq!(cal.entries[59].slot_id, 160);
    }

    #[test]
    fn leader_for_slot_is_deterministic() {
        let vs = make_validator_set(3);
        let mut tip = [0u8; 32];
        tip[0..4].copy_from_slice(&[0xde, 0xad, 0xbe, 0xef]);
        let r1 = leader_for_slot(42, tip, &vs);
        let r2 = leader_for_slot(42, tip, &vs);
        assert_eq!(r1, r2);
    }

    #[test]
    fn leader_for_slot_distributes() {
        let vs = make_validator_set(4);
        let mut counts = [0u32; 4];
        for slot in 1..=60u64 {
            let leader = leader_for_slot(slot, [0u8; 32], &vs);
            for (vi, v) in vs.iter().enumerate() {
                if &leader == v {
                    counts[vi] += 1;
                    break;
                }
            }
        }
        // Each validator should get roughly 15 of 60 slots; allow 5..35.
        for c in counts {
            assert!(c >= 5 && c <= 35, "count out of range: {c}");
        }
    }

    #[test]
    fn finalize_slot_stores_hash_and_state() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(1);
        cal.recompute(10, [0u8; 32], 0, &vs);
        let block_hash = [0xABu8; 32];
        cal.finalize_slot(11, block_hash);
        assert_eq!(cal.entries[0].state, SlotState::Finalized);
        assert_eq!(cal.entries[0].placeholder_hash, block_hash);
    }

    #[test]
    fn prune_missed_marks_overdue() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(1);
        let now_ms = 1_000_000i64;
        cal.recompute(0, [0u8; 32], now_ms, &vs);

        cal.prune_missed(now_ms + 5_000);
        assert_eq!(cal.entries[0].state, SlotState::Missed);
        assert_eq!(cal.entries[1].state, SlotState::Missed);
        assert_eq!(cal.entries[2].state, SlotState::Missed);
        assert_eq!(cal.entries[3].state, SlotState::InFlight);
        assert_eq!(cal.entries[4].state, SlotState::InFlight);
        assert_eq!(cal.entries[5].state, SlotState::Future);
    }

    #[test]
    fn get_slot_returns_none_outside_window() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(1);
        cal.recompute(100, [0u8; 32], 0, &vs);
        assert!(cal.get_slot(101).is_some());
        assert!(cal.get_slot(160).is_some());
        assert!(cal.get_slot(100).is_none());
        assert!(cal.get_slot(161).is_none());
        assert!(cal.get_slot(0).is_none());
    }

    #[test]
    fn empty_validator_set_gives_zero_leader() {
        let mut cal = SlotCalendar::new();
        cal.recompute(0, [0u8; 32], 0, &[]);
        for e in cal.entries.iter() {
            assert_eq!(e.leader_address, [0u8; 20]);
        }
    }

    #[test]
    fn ring_buffer_head_advances() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(1);
        cal.recompute(0, [0u8; 32], 0, &vs);
        assert_eq!(cal.head, 0);

        cal.finalize_slot(1, [0x01u8; 32]);
        assert_eq!(cal.head, 1);
        cal.finalize_slot(2, [0x02u8; 32]);
        assert_eq!(cal.head, 2);

        // Non-consecutive: slot 4 finalized but slot 3 (entries[2]) is still future.
        cal.finalize_slot(4, [0x04u8; 32]);
        assert_eq!(cal.head, 2);
    }

    #[test]
    fn to_json_is_parseable() {
        let mut cal = SlotCalendar::new();
        let vs = make_validator_set(2);
        cal.recompute(5, [0u8; 32], 12345, &vs);
        let json = cal.to_json();
        assert!(json.contains("\"entries\""));
        assert!(json.contains("\"slot_id\""));
        assert!(json.contains("\"state\""));
        assert!(json.contains("\"future\""));
        assert!(json.contains("\"tip_height\":5"));
        assert!(json.contains("\"computed_at_ms\":12345"));
        let parsed: serde_json::Value = serde_json::from_str(&json).expect("valid JSON");
        assert!(parsed.get("entries").is_some());
    }
}
